lib/bolt/analytics.rb

```diff
@@ -0,0 +1,141 @@
+require 'json'
+require 'locale'
+require 'logging'
+require 'securerandom'
+
+module Bolt
+  module Analytics
+    ProtocolVersion = 1
+    ApplicationName = 'bolt'
+    TrackingID = 'UA-fakefakefake-1'
+    TrackingURL = 'https://google-analytics.com/collect'
+
+    def self.build_client
+      logger = Logging.logger[self]
+
+      config_file = File.expand_path('~/.puppetlabs/bolt/analytics.yaml')
+      config = load_config(config_file)
+
+      if config['disabled'] || ENV['BOLT_DISABLE_ANALYTICS']
+        logger.debug "Analytics opt-out is set, analytics will be disabled"
+        NoopClient.new
+      else
+        if !config.key?('user-id')
+          config['user-id'] = SecureRandom.uuid
+          write_config(config_file, config)
+        end
+
+        Client.new(config['user-id'])
+      end
+    rescue StandardError => e
+      logger.debug "Failed to initialize analytics client, analytics will be disabled: #{e}"
+      NoopClient.new
+    end
+
+    def self.load_config(filename)
+      if File.exist?(filename)
+        YAML.load_file(filename)
+      else
+        {}
+      end
+    end
+
+    def self.write_config(filename, config)
+      File.write(filename, config.to_yaml)
+    end
+
+    class Client
+      attr_reader :user_id
+
+      def initialize(user_id)
+        @logger = Logging.logger[self]
+        @http = HTTPClient.new
+        @user_id = user_id
+        @executor = Concurrent.global_io_executor
+      end
+
+      def screen_view(screen)
+        screen_view_params = {
+          # Type
+          t: 'screenview',
+          # Screen Name
+          cd: screen
+        }
+
+        submit(base_params.merge(screen_view_params))
+      end
+
+      def event(category, action)
+        event_params = {
+          # Type
+          t: 'event',
+          # Event Category
+          ec: category,
+          # Event Action
+          ea: action
+        }
+
+        submit(base_params.merge(event_params))
+      end
+
+      def submit(params)
+        # Handle analytics submission in the background to avoid blocking the
+        # app or polluting the log with errors
+        Concurrent::Future.execute(executor: @executor) do
+          @logger.debug "Submitting analytics: #{JSON.pretty_generate(params)}"
+          @http.post(TrackingURL, query: params)
+          @logger.debug "Completed analytics submission"
+        end
+      end
+
+      # These parameters have terrible names. See this page for complete documentation:
+      # https://developers.google.com/analytics/devguides/collection/protocol/v1/parameters
+      def base_params
+        {
+          v: ProtocolVersion,
+          # Client ID
+          cid: @user_id,
+          # Tracking ID
+          tid: TrackingID,
+          # Application Name
+          an: ApplicationName,
+          # Application Version
+          av: Bolt::VERSION,
+          # Anonymize IPs
+          aip: true,
+          # User locale
+          ul: Locale.current.to_rfc
```

Comment:
I don't believe so.

```rb
# * Returns: an Array of the current locales (Locale::Tag's class).
def current
  unless Thread.current[:current_languages]
    loc = driver_module.locales
    Thread.current[:current_languages] = loc ? loc : Locale::TagList.new([default])
  end
  Thread.current[:current_languages]
end
```

It seems it will return a default if no other locale is specified.